far/strmix.cpp

```diff
@@ -911,71 +993,33 @@ bool SearchString(string_view const Source, const string& Str, const string& Upp
 	if (!Regexp && PreserveStyle && PreserveStyleReplaceString(Source, Str, ReplaceStr, CurPos, Case, WholeWords, WordDiv, Reverse, *SearchLength))
 		return true;
 
+	const int SourceSize = static_cast<int>(Source.size());
+
 	if (Reverse)
 	{
+		// MZK 2018-04-01 BUGBUG: regex reverse search: "^$" does not match empty string
 		Position--;
 
-		if (Position >= static_cast<int>(Source.size()))
-			Position = static_cast<int>(Source.size() - 1);
+		if (Position >= SourceSize)
+			Position = SourceSize - 1;
 
-		if (Position<0)
+		if (Position < 0)
 			return false;
 	}
 
-	if ((Position < static_cast<int>(Source.size()) || (!Position && Source.empty())) && !Str.empty())
+	if ((Position < SourceSize || (Position == 0 && Source.empty())) && !Str.empty())
 	{
 		if (Regexp)
 		{
-			intptr_t n = re.GetBracketsCount();
-			bool found = false;
-			int half = 0;
-			if (!Reverse)
-			{
-				if (re.SearchEx(Source.raw_data(), Source.raw_data() + Position, Source.raw_data() + Source.size(), pm, n, hm))
-				{
-					found = true;
-				}
-				else
-				{
-					ReMatchErrorMessage(re);
-				}
-			}
-			else
-			{
-				int pos = 0;
-				for (;;)
-				{
-					if (!re.SearchEx(Source.raw_data(), Source.raw_data() + pos, Source.raw_data() + Source.size(), pm + half, n, hm))
-					{
-						ReMatchErrorMessage(re);
-						break;
-					}
-					pos = static_cast<int>(pm[half].start);
-					if (pos > Position)
-						break;
-
-					found = true;
-					++pos;
-					half = n - half;
-				}
-				half = n - half;
-			}
-			if (found)
-			{
-				*SearchLength = pm[half].end - pm[half].start;
-				CurPos = pm[half].start;
-				ReplaceStr = ReplaceBrackets(Source, ReplaceStr, pm + half, n, hm);
-			}
-
-			return found;
+			return SearchStringRegex(Source, re, pm, hm, Position, Reverse, ReplaceStr, CurPos, SearchLength);
 		}
 
-		if (Position == static_cast<int>(Source.size()))
+		if (Position == SourceSize)
 			return false;
 
-		const int Length = *SearchLength = (int)Str.size();
+		const int Length = *SearchLength = static_cast<int>(Str.size());
 
-		for (int I=Position; (Reverse && I>=0) || (!Reverse && I < static_cast<int>(Source.size())); Reverse ? I--:I++)
+		for (int I = std::min(Position, SourceSize - Length); 0 <= I && I <= SourceSize - Length; Reverse ? I-- : I++)
```

Comment:
This was the root cause of the buffer overflow. If `I` was close to the end of the `Source`, the original code could continue matching characters of `Str` beyond the `Source`'s end.

Reply:
@MKadaner, could you please explain how exactly the initial issue can be reproduced?
It looks like this change [breaks](https://forum.farmanager.com/viewtopic.php?p=149038#p149038) the search logic, we need to fix it in a different way.

Reply:
Thank you for reverting. Sorry for not checking repeated search.

Repro: *debug* build. File contains
```
ABCD
```
Searching for `DEF` from the beginning will hit `assert` in `range::operator[]`.

Reply:
@alabuzhev, I looked at your change. The early return you added
```
	if (Haystack.empty())
		return false;
```
breaks the regex search. Now regex /^$/ does not match empty string .

I was going to fix this same problem (with reverse search only) caused by another earlier return, which was there for a long time, so you can leave this one to me as well. However, I won't be able to do this earlier than the coming weekend, and even then not sure I'll have enough time.

Reply:
Thanks, fixed. I'll leave reverse search case to you.